### PR TITLE
Use InplaceClipGradNorm for offline processing for on-device training

### PR DESCRIPTION
**Pull request by baijumeswani**
This pull request:

- Adds the CPU kernel for `InplaceClipGradNorm`
- Updates the offline tooling `ClipGradNorm` to use the new `InplaceClipGradNorm` kernel
- Updates the `Optimizer` class to work with gradients as a `TensorSeq` input.

This is an example of how the optimizer model with `AdamWOptimizer` used to look:
image

And here is the same model with the `TensorSeq` inputs for the gradientss:
image

### CODE REVIEW

File: orttraining/orttraining/python/training/onnxblock/optim/optim.py [hunk 1]
```diff
@@ -181,6 +179,7 @@ def build(self, parameters):
         params_name = "params"
         first_order_moments_name = "first_order_moments"
         second_order_moments_name = "second_order_moments"
+        gradients_name = "gradients"
         gradient_suffix = "_grad"
```
Comment (pengwa) [inline orttraining/orttraining/python/training/onnxblock/optim/optim.py:183]: pylint said "Unused variable 'gradient_suffix' (unused-variable)" :)

File: orttraining/orttraining/test/python/orttraining_test_onnxblock.py [hunk 2]
```diff
@@ -696,7 +697,7 @@ def test_grad_clipping_execution():
 
     # Prepare the onnx model with only grad clipping
     onnx_model = onnx.ModelProto()
-    onnx_model.graph.name = "AdamW Optimizer Model"
+    onnx_model.graph.name = "ClipGradNorm Model"
     onnx_model.producer_name = "grad clipping test"
```
Comment (pengwa) [inline orttraining/orttraining/test/python/orttraining_test_onnxblock.py:701]: I think we are not test, now?  ""grad clipping test"" --> "onnxblock"???

File: orttraining/orttraining/test/python/orttraining_test_onnxblock.py [hunk 3]
```diff
@@ -705,21 +706,24 @@ class GradClippingModel(onnxblock.Model):
...
-            onnx.helper.make_tensor_value_info(grad_names[-1], onnx.TensorProto.FLOAT, param.shape)
-        )
+    onnx_model.graph.input.append(
+        onnx.helper.make_tensor_sequence_value_info("gradients", onnx.TensorProto.FLOAT, None)
+    )
 
     grad_clip = GradClippingModel(2.5)
 
+    # Note the call to build() and not to __call__() as onnx does not know
+    # how to infer the output from InplaceClipGradNorm and that will result in
```
Comment (pengwa) [inline orttraining/orttraining/test/python/orttraining_test_onnxblock.py:719]: you mean, the shape infer will just throw, do you think we should have a best effort shape infer, e.g. allowing failed shape infer. 

If we define __call__() is the right function for building graph, we'd better to keep this consistency. What do you think?

File: orttraining/orttraining/test/python/orttraining_test_onnxblock.py (hunk 2 above)
Comment (pengwa) [reply orttraining/orttraining/test/python/orttraining_test_onnxblock.py:701]: oh sorry, this is in a test file. Ignore this comment pls. :)

File: orttraining/orttraining/test/training_ops/cuda/optimizer/clip_grad_norm_test.cc [hunk 4]
```diff
@@ -8,9 +8,9 @@
 namespace onnxruntime {
 namespace test {
 
-#ifdef USE_CUDA
+namespace {
```
Comment (pengwa) [inline orttraining/orttraining/test/training_ops/cuda/optimizer/clip_grad_norm_test.cc:11]: you removed "#ifdef USE_CUDA" by mistake?

File: orttraining/orttraining/training_api/optimizer.cc [hunk 5]
```diff
@@ -48,59 +49,48 @@ Status Optimizer::GenerateMomentumNamedStates() {
...
 
         // Collect second order moments and prepare for tensorseq creation
-        auto* second_order_moment_tensor = param_named_optimizer_states.at(param_name)
+        auto* second_order_moment_tensor = param_named_optimizer_states.at(name_parameter_pair.first)
                                                .momentum_named_states.at(MOMENT_STATE_NAMES[1])
                                                .GetMutable<Tensor>();
         second_order_moments.emplace_back(
             Tensor(second_order_moment_tensor->DataType(), second_order_moment_tensor->Shape(),
                    second_order_moment_tensor->MutableDataRaw(), second_order_moment_tensor->Location()));
-      } else {
```
Comment (pengwa) [inline orttraining/orttraining/training_api/optimizer.cc:96]: why remove this else branch?

**Comment by pengwa** [comment]
left some comments, will do more check later.

File: orttraining/orttraining/test/python/orttraining_test_onnxblock.py (hunk 3 above)
Comment (baijumeswani) [reply orttraining/orttraining/test/python/orttraining_test_onnxblock.py:719]: The problem is that the shape is needed to be known for outputs. So failing shape inference for outputs is not acceptable. We need to somehow populate the output shapes. Let me see if I can update the implementation to populate the shape for the output of this node.

File: orttraining/orttraining/training_api/optimizer.cc (hunk 5 above)
Comment (baijumeswani) [reply orttraining/orttraining/training_api/optimizer.cc:96]: Earlier, we were iterating over the graph inputs. So, if we encounter an unknown user input, we report it to the user in the `else` branch. We no longer need to iterate over the user inputs since the user inputs does not define any specific tensor. We need to iterate only over the named parameters and built the gradients and moments from there.

Having said that we should add a check for any unknown user inputs. Will add that.

File: orttraining/orttraining/training_ops/cpu/optimizer/clip_grad_norm/clip_grad_norm.cc [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+template <typename T>
+Status InplaceClipGradNorm<T>::Compute(OpKernelContext* ctx) const {
+  const TensorSeq* gradients = ctx->Input<TensorSeq>(0);
+
+  const T total_norm = GetL2Norm<T>(gradients);
+
+  ClipGradNorm(total_norm, max_norm_, &gradients);
+
+  // Populate the output sequence tensors.
+  AllocatorPtr alloc;
```
Comment (pengwa) [inline orttraining/orttraining/training_ops/cpu/optimizer/clip_grad_norm/clip_grad_norm.cc:71]: can we delay creating such a allocator when it is really needed?

File: orttraining/orttraining/training_ops/cpu/optimizer/clip_grad_norm/clip_grad_norm.cc [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+  T l2_norm = 0;
+  for (const auto& tensor : *gradients) {
+    l2_norm +=
+        ReduceAggregatorSumSquare<T>(tensor.Shape().Size(), *tensor.Data<T>()).aggall(tensor.Data<T>());
+  }
+  return reduce_sqrt<T>(l2_norm);
+}
+
+template <typename T>
+void ClipGradNorm(T total_norm, T max_norm, const TensorSeq** gradients) {
```
[the remark this answers is not in the document]
Comment (baijumeswani) [reply orttraining/orttraining/training_ops/cpu/optimizer/clip_grad_norm/clip_grad_norm.cc:27]: We can but we will still have to `const_cast` the tensor within the `TensorSeq`. `TensorSeq` does not have a non const iterator or a way to get reference to a non const tensor.

File: orttraining/orttraining/training_api/optimizer.cc [hunk 8]
```diff
@@ -22,12 +22,53 @@ constexpr char GROUP_ZERO_NAME[] = "group0";
...
+    "first_order_moments",
+    "second_order_moments"};
+
+template <std::size_t NumInputs>
+Status GraphInputsAreExpected(const std::vector<std::string>& actual_graph_inputs,
+                              const std::array<const char*, NumInputs>& expected_graph_inputs) {
+  const auto stringify = [](const auto& container) {
+    std::string container_str("[");
+    for (const auto& val : container) {
+      container_str += std::string(val) + std::string(", ");
```
Comment (pengwa) [inline orttraining/orttraining/training_api/optimizer.cc:39]: maybe a stringstream fit this loop better?

File: orttraining/orttraining/training_api/optimizer.cc [hunk 9]
```diff
@@ -22,12 +22,53 @@ constexpr char GROUP_ZERO_NAME[] = "group0";
...
+constexpr std::array<const char*, 6> AdamWOptimizerInputs = {
+    "learning_rate",
+    "step",
+    "params",
+    "gradients",
+    "first_order_moments",
+    "second_order_moments"};
+
+template <std::size_t NumInputs>
+Status GraphInputsAreExpected(const std::vector<std::string>& actual_graph_inputs,
```
Comment (pengwa) [inline orttraining/orttraining/training_api/optimizer.cc:34]: nit: gsl_span?

**Review by pengwa — DISMISSED** [review]

File: orttraining/orttraining/training_api/optimizer.cc [hunk 10]
```diff
@@ -22,11 +22,51 @@ constexpr char GROUP_ZERO_NAME[] = "group0";
 // TODO: Conolidate with frontend tooling
 const std::vector<std::string> MOMENT_STATE_NAMES{"momentum0", "momentum1"};
 
-constexpr char LearningRateName[] = "learning_rate";
-constexpr char StepName[] = "step";
-constexpr char ParamsName[] = "params";
-constexpr char FirstOrderMomentsName[] = "first_order_moments";
-constexpr char SecondOrderMomentsName[] = "second_order_moments";
+constexpr std::array<const char*, 6> AdamWOptimizerInputs = {
```
Comment (edgchen1) [inline orttraining/orttraining/training_api/optimizer.cc:25]:
```suggestion
constexpr std::array AdamWOptimizerInputs = {
```

File: orttraining/orttraining/training_api/optimizer.cc [hunk 11]
```diff
@@ -22,11 +22,51 @@ constexpr char GROUP_ZERO_NAME[] = "group0";
...
+    "learning_rate",
+    "step",
+    "params",
+    "gradients",
+    "first_order_moments",
+    "second_order_moments"};
+
+template <std::size_t NumInputs>
+Status GraphInputsAreExpected(const gsl::span<std::string> actual_graph_inputs,
+                              const std::array<const char*, NumInputs>& expected_graph_inputs) {
```
Comment (edgchen1) [inline orttraining/orttraining/training_api/optimizer.cc:35]: `gsl::span<const char* const>` would avoid the need for a template

File: orttraining/orttraining/training_ops/cpu/optimizer/clip_grad_norm/clip_grad_norm.cc [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+  }
+  return reduce_sqrt<T>(l2_norm);
+}
+
+template <typename T>
+void ClipGradNorm(T total_norm, T max_norm, TensorSeq& gradients) {
+  const T clip_coefficient = std::min(max_norm / (total_norm + static_cast<T>(Epsilon)), static_cast<T>(1.0f));
+
+  for (const auto& grad : gradients) {
+    auto& tensor = const_cast<Tensor&>(grad);
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/cpu/optimizer/clip_grad_norm/clip_grad_norm.cc:31]: would it be possible to update TensorSeq to allow mutable access to its tensors?

File: orttraining/orttraining/training_ops/cpu/optimizer/clip_grad_norm/clip_grad_norm.cc [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+  }
+}
+
+void PopulateOutput(OpKernelContext* ctx, const TensorSeq* gradients, TensorSeq* clipped_gradients) {
+  if (const_cast<TensorSeq*>(gradients) == clipped_gradients) {
+    return;
+  }
+
+  AllocatorPtr alloc;
+  ORT_ENFORCE(ctx->GetTempSpaceAllocator(&alloc).IsOK(), "InplaceClipGradNorm: Unable to get an allocator.");
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/cpu/optimizer/clip_grad_norm/clip_grad_norm.cc:42]: consider returning a Status from this function

File: orttraining/orttraining/training_ops/cpu/optimizer/clip_grad_norm/clip_grad_norm.cc [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+}  // namespace
+
+ONNX_OPERATOR_KERNEL_EX(
+    InplaceClipGradNorm,
+    kMSDomain,
+    1,
+    kCpuExecutionProvider,
+    (*KernelDefBuilder::Create())
+        .Alias(0, 0) /* Return updated gradients in-place */
+        .TypeConstraint("S_GRAD", DataTypeImpl::AllFixedSizeSequenceTensorTypes()),
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/cpu/optimizer/clip_grad_norm/clip_grad_norm.cc:62]: this kernel registration actually only supports float tensors right?

File: orttraining/orttraining/training_ops/cpu/optimizer/clip_grad_norm/clip_grad_norm.cc [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+    InplaceClipGradNorm<float>);
+
+template <typename T>
+Status InplaceClipGradNorm<T>::Compute(OpKernelContext* ctx) const {
+  const TensorSeq* gradients = ctx->Input<TensorSeq>(0);
+
+  const T total_norm = GetL2Norm<T>(*gradients);
+
+  auto grads = const_cast<TensorSeq*>(gradients);
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/cpu/optimizer/clip_grad_norm/clip_grad_norm.cc:71]: should we consider updating the kernel infrastructure to support in-place updates without requiring `const_cast`ing of inputs?

**Review by edgchen1 — COMMENTED** [review]
kernel def hash update looks fine. had a few other comments

File: orttraining/orttraining/training_ops/cpu/optimizer/clip_grad_norm/clip_grad_norm.cc (hunk 12 above)
Comment (baijumeswani) [reply orttraining/orttraining/training_ops/cpu/optimizer/clip_grad_norm/clip_grad_norm.cc:31]: So far, we have 2 ops in training that would benefit from this (AdamWOptimizer and InplaceClipGradNorm).
I think we probably should. I haven't taken a look at other examples yet, but I am wondering how one would go about performing operations on the tensors inside the tensorseq (other than recreate new tensors from existing ones) if the only access is non mutable.
Having said that, I think I would like to probably get to that in another PR. I can create a work item for this.

File: orttraining/orttraining/training_ops/cpu/optimizer/clip_grad_norm/clip_grad_norm.cc [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+  const T clip_coefficient = std::min(max_norm / (total_norm + static_cast<T>(Epsilon)), static_cast<T>(1.0f));
+
+  for (const auto& grad : gradients) {
+    auto& tensor = const_cast<Tensor&>(grad);
+    MakeEigenArrayMap<T>(tensor) *= clip_coefficient;
+  }
+}
+
+void PopulateOutput(OpKernelContext* ctx, const TensorSeq* gradients, TensorSeq* clipped_gradients) {
+  if (const_cast<TensorSeq*>(gradients) == clipped_gradients) {
```
[the remark this answers is not in the document]
Comment (baijumeswani) [reply orttraining/orttraining/training_ops/cpu/optimizer/clip_grad_norm/clip_grad_norm.cc:37]: We probably don't need it. Thanks.

File: orttraining/orttraining/training_ops/cpu/optimizer/clip_grad_norm/clip_grad_norm.cc (hunk 15 above)
Comment (baijumeswani) [reply orttraining/orttraining/training_ops/cpu/optimizer/clip_grad_norm/clip_grad_norm.cc:71]: Yes, this would be nice. I think this might be a training only requirement, but we have a use case for updating the inputs in place. I can create a work item for this along with the mutable access of tensors in tensorseq.

File: orttraining/orttraining/training_ops/cpu/optimizer/clip_grad_norm/clip_grad_norm.cc (hunk 14 above)
Comment (baijumeswani) [reply orttraining/orttraining/training_ops/cpu/optimizer/clip_grad_norm/clip_grad_norm.cc:62]: Yes, the kernel registration is only for float tensors. For now, this kernel only support float tensors. We can update in the future as needed.

File: orttraining/orttraining/training_api/optimizer.cc [hunk 17]
```diff
@@ -22,11 +22,50 @@ constexpr char GROUP_ZERO_NAME[] = "group0";
...
+
+    return container_str;
+  };
+
+  const auto construct_unexpected_input_status = [&stringify](const auto& actual_inputs, const auto& expected_inputs) {
+    std::ostringstream error_stream;
+    error_stream << "Invalid graph inputs."
+                 << "\n\tExpected: " << stringify(expected_inputs)
+                 << "\n\tActual: " << stringify(actual_inputs);
+    return ORT_MAKE_STATUS(StatusCategory::ONNXRUNTIME, StatusCode::INVALID_ARGUMENT, error_stream.str());
```
Comment (edgchen1) [inline orttraining/orttraining/training_api/optimizer.cc:54]:
```suggestion
    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, error_stream.str());
```

**Review by edgchen1 — APPROVED** [review]
